### PR TITLE
Build: migrate to file lists for driver invocations

### PR DESCRIPTION
**Pull request by compnerd**
When invoking the driver use file lists rather than passing all the sources on the command line.  This is particularly import for Windows where the command line limit is 32k.  With deep dependency sets and large projects, it is possible to overrun this limit and fail to invoke the command.  The driver itself will invoke the frontend using response files if the invocation will overflow the command line limit.

### CODE REVIEW

**Comment by compnerd** [comment]
It seems that changing https://github.com/apple/swift-package-manager/pull/6573/files#diff-31b280ef997c175a0c39198547ce45d473e9fb89ac71aaa5ed2ad7e4a92473c8R549 to a file list breaks the ability to build targets individually? 😕

**Comment by compnerd** [comment]
@swift-ci please smoke test

**Comment by MaxDesiatov** [comment]
@swift-ci test windows

File: Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift [hunk 1]
```diff
@@ -677,6 +676,13 @@ public final class SwiftTargetBuildDescription {
         self.buildParameters.triple.isDarwin() && self.target.type == .library
     }
 
+    private func writeFileList(includeDerivedSources: Bool = false) throws -> AbsolutePath {
```
Comment (tomerd) [inline Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift:679]: how are these files getting deleted?

File: Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift [hunk 2]
```diff
@@ -546,7 +546,8 @@ public final class SwiftTargetBuildDescription {
         }
 
         result.append("-c")
-        result.append(contentsOf: self.sources.map(\.pathString))
+        // FIXME: Eliminate side effect.
+        try result.append("@\(self.writeFileList(includeDerivedSources: true).pathString)")
```
Comment (tomerd) [inline Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift:550]: should this be conditional on the number of sources? seems like it may be easier to debug when not using a file so using it only when we _need_ could be a good strategy?

File: Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift (hunk 1 above)
Comment (tomerd) [reply Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift:679]: I think this function should take the array of sources instead of the includeDerivedSources argument, this leave control in the hand of the call site and not push the logic to this function

File: Tests/CommandsTests/BuildToolTests.swift [hunk 3]
```diff
@@ -150,7 +150,7 @@ final class BuildToolTests: CommandsTestCase {
             do {
                 let result = try build(["--product", "exec1"], packagePath: fullPath)
                 XCTAssertMatch(result.binContents, ["exec1"])
-                XCTAssertNoMatch(result.binContents, ["exec2.build"])
+                XCTAssertMatch(result.binContents, ["exec2.build"])
```
Comment (tomerd) [inline Tests/CommandsTests/BuildToolTests.swift:153]: these test changes make me worried? seems like we are also changing the logic, not just the use of a file?

**Review by tomerd — CHANGES_REQUESTED** [review]
few follow up questions

File: Tests/CommandsTests/BuildToolTests.swift (hunk 3 above)
Comment (compnerd) [reply Tests/CommandsTests/BuildToolTests.swift:153]: The logic is the same - the differences here are because `exec2.build` is a directory, and that directory will now exist because we wrote out the file list.

File: Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift (hunk 2 above)
Comment (compnerd) [reply Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift:550]: It is not really the number of sources that is determinant of the need, it is the length of the command line.  We do have https://github.com/apple/swift-driver/blob/main/Sources/SwiftDriver/Utilities/System.swift#L58-L116 that can help identify if we are exceeding the length.

I only disagree with your statement in the "may be" aspect - I would say it *is* easier to debug with the command line being complete.

File: Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift (hunk 1 above)
Comment (MaxDesiatov) [reply Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift:679]: I also hope if these files get deleted there was an option to not delete and keep them on the file system after the build for debugging purposes. Had exact same problem with `clang` today. Looks like that has a `-save-temps` flag, but apparently it respects that only for temporary `.o` files and similar temporary build artifacts. We don't have to have the same exact option, but some flag to not delete these files would be appreciated.

Comment (compnerd) [reply Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift:679]: The file is never deleted, it is emitted and kept, which I was happily surprised about.  If the files are reaped, it would mean that the command that is displayed is not possible to replicate as the file list is needed to execute the child process.

Sure, I like the idea of the taking the sources.

Comment (compnerd) [reply Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift]: @MaxDesiatov I had that same exact concern which is why I was happy that it doesn't get deleted 😄 That is exactly the reason the tests needed adjustment - the `.build` directories are now created early to store the file list.

Comment (tomerd) [reply Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift]: if they are not deleted, is there a risk of conflicts between runs? also, where exactly are they written to?

File: Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift (hunk 2 above)
Comment (tomerd) [reply Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift]: is detecting the length an API we can use, seems internal to driver

File: Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift (hunk 1 above)
Comment (compnerd) [reply Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift]: I don't think that it is any higher risk than say running a build on the same tree in a VM and on the host.  They are written to `[target].build` (which is why the tests need to be adjusted - the directory is created eagerly as the file lists are written out just like the output file map and other such files).

Comment (tomerd) [reply Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift]: so this is in the `.build` directory, but what happens when the file already exit from previous runs?

Comment (compnerd) [reply Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift]: It will be over-written.  There is no risk of things getting mixed up AFAICT.  Do you have a concrete example of where you see this risk?

**Review by neonichu — APPROVED** [review]

File: Sources/Build/LLBuildManifestBuilder.swift [hunk 4]
```diff
@@ -626,7 +626,8 @@ extension LLBuildManifestBuilder {
             tempsPath: target.tempsPath,
             objects: try target.objects,
             otherArguments: try target.compileArguments(),
-            sources: target.sources,
+            sourceCount: target.sources.count,
+            fileListPath: try target.writeFileList(target.sources),
```
Comment (tomerd) [inline Sources/Build/LLBuildManifestBuilder.swift:630]: I would prefer we make the "writeFileList" functional accessible to this module than change the signature like this

Comment (compnerd) [reply Sources/Build/LLBuildManifestBuilder.swift:630]: I'm not sure I understand what you are suggesting.  Do you want to pass `target` down and invoke `writeFileList` in the callee?  I will note that we do something similar a few lines below that I didn't change.

Comment (tomerd) [reply Sources/Build/LLBuildManifestBuilder.swift:630]: was trying to eliminate having to pass "sourceCount". is it very useful? @neonichu ?

File: Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift (hunk 1 above)
Comment (tomerd) [reply Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift]: i feel strongly that:
1. we should have a good debugging story
2. the code is easy to maintain

using files instead of list of source is hurting debugging so I would restrict that to when we _have_ to do it (eg only on Windows, or when we detect its needed), but if we feel we _have_ to do it, then I remain concerned about the files being left around and then getting mixed up in future invocation  (the side effect issue). if the code cleans it up I would feel more comfortable with this change, but that would kill debuggability, which why i am having a hard time with this change

maybe the code should make sure to delete these files before the invocation to eliminate the risk of mixups from future mistakes

File: Sources/Build/LLBuildManifestBuilder.swift (hunk 4 above)
Comment (compnerd) [reply Sources/Build/LLBuildManifestBuilder.swift:630]: Oh, I see!  Hmm, we would need to pass `target` down.  I do agree with the sentiment though - I did first try to not pass `sourceCount` down.  That is used entirely for display purposes though, not building.

File: Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift [hunk 5]
```diff
@@ -689,6 +688,13 @@ public final class SwiftTargetBuildDescription {
         self.buildParameters.triple.isDarwin() && self.target.type == .library
     }
 
+    func writeFileList(_ sources: [AbsolutePath], _ variant: String = "") throws -> AbsolutePath {
```
Comment (tomerd) [inline Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift:691]: can we call this writeSourceFileList

File: Sources/Build/LLBuildManifestBuilder.swift (hunk 4 above)
Comment (tomerd) [inline Sources/Build/LLBuildManifestBuilder.swift:630]: also can we call this sourceListPath or something that make it clear its the sources.

File: Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift (hunk 1 above)
Comment (compnerd) [reply Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift]: Can you explain _how_ this is hurting debugging?  I don't think that this actually hurts debugging in practice.  Deleting the file before invocation would not work - it would not be available when needed (when compiling the code).

Comment (neonichu) [reply Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift]: I think the real solution to that is having the file lists be generated during the build, then we would get deletion automatically through stale-file removal. That does seem separate from this PR, but we could do it as a pre-requisite.

Re: debugging, it seems like in `swift build --verbose`, we get both driver and individual frontend invocations. The frontend invocations are showing the concrete files verbatim even when using a file list, so that seems fine to me.

File: Sources/Build/LLBuildManifestBuilder.swift (hunk 4 above)
Comment (neonichu) [reply Sources/Build/LLBuildManifestBuilder.swift:630]: Don't think it is super useful, would be fine to remove it. I only put it there as I was trying to match the existing swift-compiler-tool behavior, but I don't think we have to do that going forward.

File: Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift (hunk 1 above)
Comment (tomerd) [reply Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift]: the point I am trying to make is that we have a situation where code like `writeOutputFileMap` relies on side effects (files on disk). this is risky and I dont think we should add more of this kind of code (as this PR does), we should instead figure out a way to eliminate such side effect usage

obviously, we cannot simply delete these files at the end of the build because we need them to enable debugging, ie if you want to invoke the command manually

basically what I would like to see these files are purged before you invoke “swift build” again, to eliminate the side-effect risk

since this PR does not offer a way to deal with the side-effect risk, I was trying to come up with suggestions on how to use this only when necessary, so at least we reduce the chance of the risk manifesting itself. maybe @neonichu has ideas how to deal with it better

File: Sources/Build/LLBuildManifestBuilder.swift (hunk 4 above)
Comment (compnerd) [reply Sources/Build/LLBuildManifestBuilder.swift:630]: Sure, though there is no guarantee that the file list is the sources only.  We can have a file list for libraries as well.

File: Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift (hunk 1 above)
Comment (compnerd) [reply Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift]: The files are overwritten each time the build is planned, so it is in effect purging the file - it just overwrites it.  I would agree if the name of the file were not deterministic.

Comment (neonichu) [reply Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift]: We don't have a way to purge previous files outside of what is tracked by the build system, because we cannot know which files were exactly generated during another invocation. I don't think we should invent a new mechanism to do so, so the only viable solution seems to be to generate these files during the build which I believe was always the intent to eventually do, then the deletion comes for free from stale-file removal as I mentioned.

Comment (compnerd) [reply Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift]: @neonichu that captures exactly my sentiments as well.

Comment (tomerd) [reply Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift]: > don't think we should invent a new mechanism to do so, so the only viable solution seems to be to generate these files during the build which I believe was always the intent to eventually do, then the deletion comes for free from stale-file removal as I mentioned.

that sounds good to me

> To me that seems like something that is separate from this PR, though. I don't think we're increasing the risk of anything like this happening because already every target will have a link file list and every compilation command will have an output file map, so already every single command is subject to this.

still uncomfortable adding more fuel to the fire. how difficult is achieving the above ^^? if it is within reach maybe we can do that for `writeOutputFileMap` first then use that technique in this PR as well

Comment (tomerd) [reply Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift]: @neonichu how about we explore that direction a bit. if a heavy lift, we can merge and track as an open issue. otherwise, lets fix the architecture and not add more floors on shaky foundation

Comment (compnerd) [reply Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift]: @tomerd one thing to note - this is *blocking* SPM usage on windows.  We cannot build parts of the Swift project even (e.g. DocC).  Can you help get CMake support integrated into DocC in the mean time?

File: Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift (hunk 2 above)
Comment (tomerd) [reply Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift]: @compnerd I am happy to take this change in its current form if it is put behind `#if windows`

File: Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift (hunk 1 above)
Comment (tomerd) [reply Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift]: @compnerd I am happy to take this change in its current form if it is put behind #if windows

Comment (compnerd) [reply Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift]: @tomerd how do we ensure that regressions are not introduced for that path?  Could we add an explicit manifest for the tests and run tests via CMake?

Comment (tomerd) [reply Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift]: not sure how this is different from the rest of SwiftPM in that regard

in any case, i think this is the best compromise we can for now, and we should def make it the default once we have a systematic way to handle the side effects which @neonichu is looking at

Comment (neonichu) [reply Sources/Build/BuildDescription/SwiftTargetBuildDescription.swift]: I think it would be incorrect to put this change behind `#if windows` tbh, there's nothing Windows specific here. We can maybe live with it as a temporary thing, but it'll increase the code complexity significantly for no reason which seems worse to me than taking the theoretical risk of stale files.

**Comment by compnerd** [comment]
@swift-ci please smoke test

**Review by neonichu — APPROVED** [review]

**Review by tomerd — APPROVED** [review]

**Comment by neonichu** [comment]
The failures in self-hosted look legitimate, these jobs run some additional integration tests which are in separate package. Easiest way to repro locally is using `Utilities/build-using-self`.

**Comment by compnerd** [comment]
@swift-ci please test

**Comment by compnerd** [comment]
@swift-ci please smoke test
